roofit/roofitcore/src/RooProdPdf.cxx

```diff
@@ -1911,15 +1911,55 @@ RooArgSet* RooProdPdf::findPdfNSet(RooAbsPdf const& pdf) const
 }
 
 
+
+/// Add some full PDFs to the factors of this RooProdPdf.
+void RooProdPdf::addPdfs(RooArgSet const& pdfs)
+{
+   size_t numExtended = (_extendedIndex==-1) ? 0 : 1;
+
+   for(auto arg : pdfs) {
+      RooAbsPdf* pdf = dynamic_cast<RooAbsPdf*>(arg);
+      if (!pdf) {
+         continue;
+      }
+      if(pdf->canBeExtended()) {
+         if (_extendedIndex == -1) {
+            _extendedIndex = _pdfList.size();
+         } else {
+            numExtended++;
+         }
+      }
+      _pdfList.add(*pdf);
+      _pdfNSetList.emplace_back(std::make_unique<RooArgSet>("nset"));
+   }
+
+   // Protect against multiple extended terms
+   if (numExtended>1) {
+      coutW(InputArguments) << "RooProdPdf::addPdfs(" << GetName()
+                            << ") WARNING: multiple components with extended terms detected,"
+                            << " product will not be extendible." << endl ;
+      _extendedIndex = -1 ;
+   }
+
+   // Reset cache
+   _cacheMgr.reset() ;
+
+}
+
 /// Remove some PDFs from the factors of this RooProdPdf.
-/// Private implementation detail.
 void RooProdPdf::removePdfs(RooArgSet const& pdfs)
 {
   // Remember what the extended PDF is
   RooAbsArg const* extPdf = _extendedIndex >= 0 ? &_pdfList[_extendedIndex] : nullptr;
 
-  // Actually remove the PDFs
-  _pdfList.remove(pdfs);
+  // Actually remove the PDFs and associated nsets
```

Comment:
I remember that I have added this `removePdfs()` function recently. Thanks for finding that bug!